lib/ecto/repo/schema.ex

```diff
@@ -78,31 +78,69 @@ defmodule Ecto.Repo.Schema do
   defp extract_header_and_fields(rows, schema, dumper, autogen_id, adapter) do
     mapper = init_mapper(schema, dumper, adapter)
 
-    Enum.map_reduce(rows, %{}, fn fields, header ->
-      {fields, header} = Enum.map_reduce(fields, header, mapper)
```

Comment:
@josevalim 

> What is a map? :) I think we normalize all of the headears and fields before sending them to the adapter, right?

Sorry for not being clear. I meant the `header` in this context. :)

The problem I encountered was that since we don't process items in each row the same order as when we [generate SQL](https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/adapters/sql.ex#L509-L518), the query params counter in this step might not be in the same as when the adapter processes the params.

_I might be missing something very obvious (which happens very usually) so please bear with me._ :)

Reply:
It is expected that adapters will process the parameters in the order given in headers. We can even explicitly document it. Would it address this issue? :)

Reply:
Here is a suggestion to try to simplify this:

1. First we do a traversal, exactly like we do today, although we also check if any of the fields contain an ecto query

2. If any of the fields contain an ecto query, then we do another traversal. Except that on this new traversal we already have a list of headers and therefore we have the expected ordering. So now we can build the queries with the relevant counters.

This should make the common case as efficient as it is today and it should simplify the work we have to do when we do have a query. Thoughts?

Reply:
Oh that indeed sounds better, I will give that a try later today 👍. Thanks for the help!